### PR TITLE
Convert redis store into redis hook

### LINKED ISSUE #159: Refactoring handlers to accomodate hooks, and events and stores
A discussion with the crew.

For CRUD operations, we would like the following "flow" within handlers.
 
## Creation and Update

  1. Realize the object from the inputs, using the existing "realize" functions.
    1. The realization function produced our initial, default StoredObject
    2. Call our "Realization" hooks, which may return a modified StoredObject
        - these are the before-create/update/delete hooks
  2. Wrap the StoredObject in the appropriate Event.
  3. Call the Event hooks, which cannot transform, and must all run succesfully (kafka feed...)
       - these are the after-create/update/delete-hooks, except those functions dont take events, they take the object itself.  This is a deviation
  4. Call the stores, passing in the StoredObject, to perform the operation, it must return succesfully
  5. Pass the Event (not just the StoredObject) on the core.async channel

## Deletion
  1. Replicate the realize- flow with a destroy- flow
  2. Product the Event, and it should contain the model, which is an optional field in the schema
  3. Call the Event hooks, which cannot transform, and must all run succesfully (kafka feed...)
  4. Call the stores delete/destroy, it must be successful
  5. Pass the Event on the core.async channel

## Realization Hooks
For every Stored Object type, we have a vector of Hook objects, modeled on GUNDAM hook.  Hooks are all "required" in that they must all return succesfully.  They must be run in a specific order, this allows the specification of hook order as a means of control flow configuration. Hooks can return a modified realized object.  If they return nil the current realized object is used.  They take the same parameter set as the realize functions.

## Commentary
Craig: Note that we isolate business logic in the realize and detroy steps, from the implementation logic in the Event hooks.

Yann: The hooks seem similiar to ring middlware, but middleware is defined on a per handler basis, and is dealing with non-realized input parameters (the request map).

### PR DESCRIPTION
**Pull request by polygloton**
Relates to #255
Closes #159 

- Implements redis event producer as a hook
- Wires-up ES event producer as hook
- Refactors flow hook code
- Removes dead code
- Needs issue #271 as post merge clean up

@quoll and I paired on this PR.  :)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -30,4 +30,4 @@ env:
       - CTIA_STORE_ES_INDEXNAME=elasticsearch
       - CTIA_PRODUCER_ES_URI=http://127.0.0.1:9200
```
Comment (gbuisson) [inline .travis.yml]: do you think we should discard this and use CTIA_HOOK_ES instead ?

**Comment by gbuisson** [comment]
this is a lot of work! sounds good to me !

**Comment by yogsototh** [comment]
I really like it! 👍

File: .travis.yml (hunk 1 above)
Comment (polygloton) [inline .travis.yml]: Will do.

File: src/ctia/flows/autoload.clj [hunk 2]
```diff
@@ -46,7 +46,11 @@
...
-                                                (finally
-                                                  nil))))))
+  ([]
+   (load-hooks!
+    (:hooks-classes (p/read))))
+  ([profiles :- [s/Keyword]]
+   (load-hooks!
+    (:hooks-classes (try (p/read "project.clj" profiles)
+                         (finally
+                           nil))))))
```
Comment (quoll) [inline src/ctia/flows/autoload.clj]: I realize that this was already here, but is it really wise to be reading configuration information from the build description file? I'm happy to do this in testing, but in production it is causing the conflation of the role of the build file.

Comment (polygloton) [inline src/ctia/flows/autoload.clj]: You are correct, it must go.  I'm actually working on a refactor for the autoloader and will submit it in a different PR, maybe tomorrow.
